python/python/lance/ray/sink.py

```diff
@@ -74,59 +165,66 @@ def write(
         blocks: Iterable[Union[pa.Table, "pd.DataFrame"]],
         ctx,
     ):
-        from ..dependencies import _PANDAS_AVAILABLE
-        from ..dependencies import pandas as pd
-
-        if self.schema is None:
-            first = next(blocks)
-            if _PANDAS_AVAILABLE and isinstance(first, pd.DataFrame):
-                schema = pa.Schema.from_pandas(first)
-            else:
-                schema = first.schema
-            blocks = chain([first], blocks)
-        else:
-            schema = self.schema
-
-        # TODO: use v2 writer later.
-        def record_batch_converter():
-            for block in blocks:
-                if _PANDAS_AVAILABLE and isinstance(block, pd.DataFrame):
-                    tbl = pa.Table.from_pandas(block, schema=schema)
-                else:
-                    tbl = block
-                yield from tbl.to_batches()
-
-        reader = pa.RecordBatchReader.from_batches(schema, record_batch_converter())
-        fragment = LanceFragment.create(self.uri, reader, fragment_id=0, schema=schema)
-        ret_schema = None if ctx.task_idx != 0 else schema
-        return (fragment, ret_schema)
+        fragment, schema = _write_fragment(blocks, self.uri, self.schema)
+        return [(cloudpickle.dumps(fragment), cloudpickle.dumps(schema))]
 
-    def on_write_start(self):
-        if self.mode == "append":
-            ds = lance.LanceDataset(self.uri)
-            self.read_version = ds.version
-            if self.schema is None:
-                self.schema = ds.schema
 
-    def on_write_complete(
-        self, write_results: List[Tuple[FragmentMetadata, Optional[pa.Schema]]]
+class LanceFragmentWriter:
+    """Write a fragment to one of Lance fragment.
+
+    This Writer can be used in case to write large-than-memory data to lance,
+    in distributed fashion.
+    """
+
+    def __init__(
+        self,
+        base_uri: str,
+        transform: Callable[[pa.Table], Union[pa.Table, Generator]] = lambda x: x,
+        schema: Optional[pa.Schema] = None,
     ):
-        logging.info(
-            "%s.on_write_complete: committing %s fragments", self, len(write_results)
+        self.base_uri = base_uri
+        self.schema = schema
+        self.transform = transform
+
+    def __call__(self, batch: Union[pa.Table, "pd.DataFrame"]) -> Dict[str, Any]:
+        """Write a Batch to the Lance fragment."""
+
+        transformed = self.transform(batch)
+        if not isinstance(transformed, Generator):
+            transformed = (t for t in [transformed])
+
+        fragment, schema = _write_fragment(transformed, self.base_uri, self.schema)
+        return pa.Table.from_pydict(
+            {
+                "fragment": [cloudpickle.dumps(fragment)],
+                "schema": [cloudpickle.dumps(schema)],
```

Comment:
Why `cloudpickle`? These both are pickle-able?

Reply:
Ray treats this function returns a `pa.Table`. Pickle them to make this schema and fragment to be stored as two columns.

Reply:
Yeah but I mean why not use just normal `pickle` from the standard library?